src/str.c

```diff
@@ -96,6 +96,80 @@ bool mg_span(struct mg_str s, struct mg_str *a, struct mg_str *b, char sep) {
   }
 }
 
+size_t mg_str_to_num_(struct mg_str str, int base, void *val, size_t val_len);
+size_t mg_str_to_num_(struct mg_str str, int base, void *val, size_t val_len) {
+  size_t i = 0, ndigits = 0;
+  uint64_t max = val_len == 1   ? 0xFF
+                 : val_len == 2 ? 0xFFFF
+                 : val_len == 4 ? 0xFFFFFFFF
+                                : (uint64_t) ~0;
+  uint64_t max2 = max / 10, result = 0;
+  while (i < str.len && (str.buf[i] == ' ' || str.buf[i] == '\t')) i++;
+  if (base == 0 && str.len >= 2) {
+    if (str.buf[i] == '0') {
+      i++;
+      base = str.buf[i] == 'b' ? 2 : str.buf[i] == 'x' ? 16 : 10;
+      if (base != 10) ++i;
+    } else {
+      base = 10;
+    }
+  }
+  switch (base) {
+    case 2:
+      while (i < str.len && (str.buf[i] == '0' || str.buf[i] == '1')) {
+        uint64_t digit = (uint64_t) (str.buf[i] - '0');
+        if (result > max2) return 0;  // Overflow
+        result *= 2;
+        if (result > max - digit) return 0;  // Overflow
+        result += digit;
+        i++, ndigits++;
+      }
+      break;
+    case 10:
+      while (i < str.len && str.buf[i] >= '0' && str.buf[i] <= '9') {
+        uint64_t digit = (uint64_t) (str.buf[i] - '0');
+        if (result > max2) return 0;  // Overflow
+        result *= 10;
+        if (result > max - digit) return 0;  // Overflow
+        result += digit;
+        i++, ndigits++;
+      }
+      break;
+    case 16:
+      while (i < str.len) {
+        char c = str.buf[i];
+        uint64_t digit = (c >= '0' && c <= '9')   ? (uint64_t) (c - '0')
+                         : (c >= 'A' && c <= 'F') ? (uint64_t) (c - '7')
+                         : (c >= 'a' && c <= 'f') ? (uint64_t) (c - 'W')
+                                                  : (uint64_t) ~0;
+        if (digit == (uint64_t) ~0) break;
+        if (result > max2) return 0;  // Overflow
+        result *= 16;
+        if (result > max - digit) return 0;  // Overflow
+        result += digit;
+        i++, ndigits++;
+      }
+      break;
+    default:
+      return 0;
+  }
+  if (ndigits == 0) return 0;  // #2322: Content-Length = 1 * DIGIT
+  if (val_len == 1) {
+    *((uint8_t *) val) = (uint8_t) result;
+  } else if (val_len == 2) {
+    *((uint16_t *) val) = (uint16_t) result;
+  } else if (val_len == 4) {
+    *((uint32_t *) val) = (uint32_t) result;
+  } else {
+    *((uint64_t *) val) = (uint64_t) result;
```

Comment:
I suggest to explicitly check for a valid val_len:

```c
else if (val_len == 8) {
} else {
 return false
}
```

Reply:
Addressed at top of function to avoid lots of work in vain